compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -620,7 +620,42 @@ pub fn orphan_check(tcx: TyCtxt<'_>, impl_def_id: DefId) -> Result<(), OrphanChe
         return Ok(());
     }
 
-    orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |ty| Ok(ty)).unwrap()
+    orphan_check_trait_ref::<!>(trait_ref, InCrate::Local, |ty| {
+        let ty::Alias(..) = ty.kind() else { return Ok(ty) };
```

Comment:
Ah, wait this doesn't look correct. Consider `Adt<AliasTy<T>>`. I think this should be:

```suggestion
        if !ty.has_projections() else {
            return Ok(ty)
        };
```

Reply:
it's correct, because the orphan check is only shallow, so with `Adt` we first check `Adt`, and if `Adt` is fundamental, we then recurse into `Adt` at which case we call this closure again (and normalize this time)